app/controllers/onboarding_controller.rb

```diff
@@ -24,6 +27,31 @@ def create
 
   def success; end
 
+  # rubocop:disable Metrics/AbcSize
+  def telegram_auth
+    check_hash = telegram_auth_params[:hash]
+    auth_data = telegram_auth_params.slice(:auth_date, :first_name, :id, :last_name)
+    check_string = auth_data.to_unsafe_h.map { |k, v| "#{k}=#{v}" }.sort.join("\n")
+
+    secret_key = OpenSSL::Digest::SHA256.new.digest(ENV['TELEGRAM_BOT_API_KEY'])
+    hash = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha256'), secret_key, check_string)
+
+    raise ActionController::BadRequest if hash.casecmp(check_hash) != 0
+    raise ActionController::BadRequest if Time.now.to_i - telegram_auth_params[:auth_date].to_i > 86_400
+
+    @user = User.find_or_create_by(
+      telegram_id: telegram_auth_params[:id],
+      first_name: telegram_auth_params[:first_name],
+      last_name: telegram_auth_params[:last_name]
+    )
+    if @user.save!
+      render json: { message: 'Success' }, status: :ok
+    else
```

Comment:
How could we reach the `else` branch here?

Reply:
> > Note: this only works over https
> 
> I was wondering if using the authorization callback instead of letting Telegram redirect might make the integration easier:
> 
>     * Would we still need to expose/tunnel the app during development? (If not, I’d prefer that as we wouldn’t need to change the dev setup.)
> 
>     * Would we still need to configure a domain for the Telegram bot? (If not, this would reduce complexity when setting up a new app instance and bot for a new project.)
> 
> 
> We could listen for the authorization callback and then manually redirect to the respective backend route. So backend-wise, nothing should change, only like a few lines of JavaScript would be required to handle the callback and redirect.
> 
> <img alt="Screen Shot 2020-10-20 at 12 06 01" width="646" src="https://user-images.githubusercontent.com/1512805/96572172-cef93900-12cc-11eb-9a86-e0892e74feea.png">

So, @tillprochaska having a look at this, we would still need to configure the telegram to `/setdomain`, if you look at the instructions on the [Linking your domain to the bot](https://core.telegram.org/widgets/login#linking-your-domain-to-the-bot), you will see that these instructions

> Once you have chosen a bot, send the /setdomain command to @Botfather to link your website's domain to the bot. Then configure your widget below and embed the code on your website.

are not specifically for the redirect method. I checked it out for the callback method as well, and it only works when you have configured an `https` domain. We could still use the callback method, if we wanted to handle the [Checking authorization](https://core.telegram.org/widgets/login#linking-your-domain-to-the-bot), but I think we might not want to do that in JS because I'm not sure how we can organize this logic into a stimulus controller, we would need to use inline JS script tags, or maybe you know how else we could do it?

Oh, re-reading your comment, you were just suggesting using the JS to redirect to the rails endpoint, as a workaround to the domain registration, right? I think that's not possible, and we have a way to redirect provided by telegram, so it makes sense to just favor that.

Reply:
I mean I haven't thought this through too much cause I am still working on the specifics, but I can imagine if the user is saved we render the success page and if they are not saved, we could render the error page to indicate they might need to get in contact with us?

I'm not sure if that answers your question, or if you are trying to imply something? The thing is that this feature is actually a login feature that we are trying to re-purpose as a authorize me to send you messages and create an account for you in our database, right? So, I think we'd need to use `find_or_create_by`, since the user could already exist? Not sure about this though. Ideally, when this is deployed there will be single use invite urls so that once a user has clicked on the link, they would not be able to access the page again, but would it be possible that a user already exists in our database when they use the link for the first time?

Or maybe you were just pointing out that by using `save!` instead of `save`, it would throw an exception if there is an issue with save and would then never reach the else clause? 

Are you being cryptic on purpose @roschaefer :stuck_out_tongue:

Reply:
> Or maybe you were just pointing out that by using `save!` instead of `save`, it would throw an exception if there is an issue with save and would then never reach the else clause?

Yes

> Are you being cryptic on purpose @roschaefer

No, I think I was assuming that `find_or_create_by` would always return a valid user record, so `@user.save` can never get `false`.

Reply:
true statement @roschaefer